practice/react-app/src/components/Comments.tsx

```diff
@@ -0,0 +1,96 @@
+import axios from 'axios';
+import { useMutation } from '@tanstack/react-query';
+import { useComments } from '../hooks/useComments';
+import { useEffect, useState } from 'react';
+import { Placeholder } from './Placeholder';
+
+const COMMENTS_API = 'http://localhost:3000/api/data/comments';
+
+export const Comments = () => {
+  const [newComment, setNewComment] = useState('');
+  const [newRating, setNewRating] = useState('');
+  const mutation = useMutation({
+    mutationFn: (comment) => {
+      return axios.post(COMMENTS_API, comment);
+    },
+    onMutate: (variables) => {
+      console.log('onMutate', variables);
+      return { variables };
+    },
+    onError: (error, variables) => {
+      console.error(error);
+      setComments(comments.filter((comment) => comment.id !== variables.id));
+    },
+    onSuccess: (data) => {
+      setComments([...comments, JSON.parse(data.config.data)]);
+    },
+  });
+  const { isLoading, data } = useComments(COMMENTS_API);
+  const [comments, setComments] = useState(data ?? []);
+
+  const handleSubmitNewComment = (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+
+    mutation.mutate({
+      id: comments.length + 1,
+      author: 'Konrad Leśniak',
+      text: newComment,
+      rating: parseInt(newRating, 10),
+    });
+
+    setNewComment('');
+    setNewRating('');
+  };
+
+  useEffect(() => {
```

Comment:
nie obędzie się bez tego useEffectu?

Reply:
No właśnie chcąc zrealizować optimistic update użyłem stanu, data po pierwszym fetchu rq był zapisywany jako undefined, więc tylko w taki sposób udało mi się to poprawnie zapisać, teraz tak myślę czy refetch z rq by tu nie pomógł